### PR TITLE
DEV: Modernize admin user fields

### PR DESCRIPTION
**Pull request by Drenmi**
### What is this change?

This PR modernizes the user fields area of the admin UI. It is largely based on the work on the emoji section. This involves:

- Lift-and-shift and glimmerification of existing code.
- Update table styles to UI guidelines.
- Add toasts for successful actions.
- Add dedicated routes for resource actions.
- Add an empty state for listings without any custom user fields.

### Verification

The following works fine after the change:

- [x] Add new user field.
- [x] Edit existing user field.
- [x] Reorder user fields.
- [x] Destroy user field.

### CODE REVIEW

File: app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import UserField from "admin/models/user-field";
+
+export default class AdminConfigAreasUserFieldsList extends Component {
+  @service dialog;
+  @service store;
+  @service adminUserFields;
+
+  get fieldTypes() {
+    return UserField.fieldTypes();
+  }
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs:17]: if this is not supposed to be dynamic, you can just do:

fields = ...;

File: app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      const prev = this.sortedFields.objectAt(idx - 1);
+      const prevPos = prev.get("position");
+
+      prev.update({ position: f.get("position") });
+      f.update({ position: prevPos });
+    }
+  }
+
+  @action
+  moveDown(f) {
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs:40]: nitpick but I hate one letter variables :p

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+      <form.Actions>
+        <form.Submit
+          class="save"
+          @icon="check"
+          @label="admin.user_fields.save"
+        />
+        <form.Button
+          @action={{this.cancel}}
+          @label="admin.user_fields.cancel"
+          class="btn-default"
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:268]: is default needed here?

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        didConfirm: () => resolve(true),
+      });
+    });
+  }
+
+  <template>
+    <Form
+      @data={{this.formData}}
+      @onSubmit={{this.save}}
+      {{didInsert this._focusName}}
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:126]: Wonder if we should try to focus the first non disabled field by default in form kit?

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+  @tracked
+  editableDisabled =
+    this.args.userField?.requirement === "for_all_users";
+  userField;
+  originalRequirement;
+
+  constructor() {
+    super(...arguments);
+    this.originalRequirement = this.args.userField.requirement;
+  }
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:31]: you can probably just do:

```
originalRequirement = this.args.userField.requirement;
```

no need to put this in constructor

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    try {
+      const isNew = this.args.userField.isNew;
+
+      await this.args.userField.save(data);
+
+      if (this.isDestroying || this.isDestroyed) {
+        return;
+      }
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:87]: DO you know why this is necessary?

File: app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      await field.destroyRecord();
+      this.fields.removeObject(field);
+    } catch (error) {
+      popupAjaxError(error);
+    }
+  }
+
+  <template>
+    {{#if this.fields}}
+      {{#each this.sortedFields as |uf|}}
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs:72]: I also don't very much like 2 letters variable when we can avoid it :D

**Comment by jjaffeux** [comment]
That's really clean work Ted :clap:

File: app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs (hunk 7 above)
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-config-areas/user-fields-list.gjs:72]: Totally agree! They almost never make things better these 1 or 2 letter vars

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs (hunk 4 above)
Comment (Drenmi) [reply app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:125]: That's an interesting idea. 💡  I'd love this behind a switch, e.g.: `@autoFocus={{true}}`.

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs (hunk 6 above)
Comment (Drenmi) [reply app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:86]: Probably not here at least. Seems like this should either stay in the listing component or be removed. I'll investigate. 👍

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs (hunk 5 above)
Comment (Drenmi) [reply app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:30]: I think you're right. This was left here after a hackier implementation. 😬

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs (hunk 3 above)
Comment (Drenmi) [reply app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:267]: It is not.

**Review by martin-brennan — COMMENTED** [review]
Gotta change the `admin/customize/user_fields/7` route to `admin/customize/user_fields/7/edit` too :P 

Can you change these flag descriptions to not all be lowercase?

https://github.com/discourse/discourse/blob/3084409ca41a3c67547379638429441cb6ea3b4b/config/locales/client.en.yml#L6981-L6996

It would be good to identify which site settings we could put into a `user_fields` area so we can show them in a  Settings tab here.

File: app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs (hunk 4 above)
Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/components/admin-user-fields-form.gjs:116]: Yes but I kinda feel like it should be the default, so you would do `@autoFocus={{false}}` if needed, but generally spekaing this is the behavior you want in a form

**Review by martin-brennan — APPROVED** [review]
Looks great Ted! After merging can you please update https://meta.discourse.org/t/creating-and-configuring-custom-user-fields/113192 with the new screenshots? I used this as the "Learn more..." link for the page.

**Comment by Drenmi** [comment]
> Looks great Ted! After merging can you please update https://meta.discourse.org/t/creating-and-configuring-custom-user-fields/113192 with the new screenshots? I used this as the "Learn more..." link for the page.

Updated!

**Pull request merged**